clippy_lints/src/loops/manual_find.rs

```diff
@@ -0,0 +1,145 @@
+use super::utils::make_iterator_snippet;
+use super::MANUAL_FIND;
+use clippy_utils::{
+    diagnostics::span_lint_and_sugg, higher, is_lang_ctor, path_res, peel_blocks_with_stmt,
+    source::snippet_with_applicability, ty::implements_trait, visitors::is_local_used,
+};
+use if_chain::if_chain;
+use rustc_errors::Applicability;
+use rustc_hir::{
+    def::Res,
+    lang_items::LangItem::{OptionNone, OptionSome},
+    BindingAnnotation, Expr, ExprKind, HirId, Node, Pat, PatKind, StmtKind,
+};
+use rustc_lint::LateContext;
+use rustc_span::source_map::Span;
+
+pub(super) fn check<'tcx>(
+    cx: &LateContext<'tcx>,
+    pat: &'tcx Pat<'_>,
+    arg: &'tcx Expr<'_>,
+    body: &'tcx Expr<'_>,
+    span: Span,
+    expr: &'tcx Expr<'_>,
+) {
+    let inner_expr = peel_blocks_with_stmt(body);
+    // Check for the specific case that the result is returned and optimize suggestion for that (more
+    // cases can be added later)
+    if_chain! {
+        if let Some(higher::If { cond, then, r#else: None, }) = higher::If::hir(inner_expr);
+        if let Some(binding_id) = get_binding(pat);
+        if let ExprKind::Block(block, _) = then.kind;
+        if let [stmt] = block.stmts;
+        if let StmtKind::Semi(semi) = stmt.kind;
+        if let ExprKind::Ret(Some(ret_value)) = semi.kind;
+        if let ExprKind::Call(Expr { kind: ExprKind::Path(ctor), .. }, [inner_ret]) = ret_value.kind;
+        if is_lang_ctor(cx, ctor, OptionSome);
+        if path_res(cx, inner_ret) == Res::Local(binding_id);
+        if let Some(last_ret) = last_ret(cx, expr);
+        then {
+            let mut applicability = Applicability::MachineApplicable;
+            let mut snippet = make_iterator_snippet(cx, arg, &mut applicability);
+            let is_ref_to_binding =
+                matches!(pat.kind, PatKind::Ref(inner, _) if matches!(inner.kind, PatKind::Binding(..)));
+            if !(matches!(pat.kind, PatKind::Binding(..)) || is_ref_to_binding) {
+                snippet.push_str(
+                    &format!(
+                        ".map(|{}| {})",
+                        snippet_with_applicability(cx, pat.span, "..", &mut applicability),
+                        snippet_with_applicability(cx, inner_ret.span, "..", &mut applicability),
+                    )[..],
+                );
+            }
+            let mut needs_copy = false;
+            match cond.kind {
+                ExprKind::Binary(_, l, r) => {
+                    for side in [l, r] {
+                        if !matches!(
+                            side.kind,
+                            ExprKind::MethodCall(_, [caller, ..], _) if is_local_used(cx, caller, binding_id)
+                        ) && is_local_used(cx, side, binding_id)
```

Comment:
This code seems to deal with a really specific edge cases. Can you add some comments what those edge cases are, please?

Reply:
I believe this `match cond.kind` can be removed now, it only works for `Copy` types but if the type is `Copy` we should add the `&` unconditionally